### PR TITLE
AP_Periph: add Toshiba LED

### PR DESCRIPTION
**Pull request by davidbitton**
This pull request adds the necessary code to drive a Toshiba LED typically integrated in to GPS modules. The code is lifted from `ToshibaLED_I2C.cpp` because AP_Notify is unavailable to f103 targets.

### CODE REVIEW

**Comment by magicrub** [comment]
code-wise I think it's OK. Good job! However, we're kinda picky on the commit msg due to multi-vehicle release cycles happening in the same repo. With nicely named commit messages its much easier to do changelogs and cherry-picks for hot-fixes for different vehicle branches. See https://ardupilot.org/dev/docs/contributing.html

TR;DR
commits msg should start with the library directory name and each commit should be limited to the scope of that library


As a courtesy, I fix that for you. Do you want to handle it or let me do it?

File: libraries/AP_HAL_ChibiOS/hwdef/f103-GPS/hwdef.dat [hunk 1]
```diff
@@ -15,4 +15,5 @@ define HAL_UART_MIN_TX_SIZE 256
 # GPS+MAG
 define HAL_PERIPH_ENABLE_GPS
 define HAL_PERIPH_ENABLE_MAG
-define HAL_PERIPH_ENABLE_NCP5623_LED
+#define HAL_PERIPH_ENABLE_NCP5623_LED
```
Comment (magicrub) [inline libraries/AP_HAL_ChibiOS/hwdef/f103-GPS/hwdef.dat:18]: lets keep HAL_PERIPH_ENABLE_NCP5623_LED defined for now

File: Tools/AP_Periph/can.cpp [hunk 2]
```diff
@@ -583,6 +583,32 @@ static void set_rgb_led(uint8_t red, uint8_t green, uint8_t blue)
         dev->transfer(&v, 1, nullptr, 0);
     }
 #endif // HAL_PERIPH_ENABLE_NCP5623_BGR_LED
+#ifdef HAL_PERIPH_ENABLE_TOSHIBA_LED
+{
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:587]: these { } scope brackets need to be indented

File: Tools/AP_Periph/can.cpp [hunk 3]
```diff
@@ -583,6 +583,32 @@ static void set_rgb_led(uint8_t red, uint8_t green, uint8_t blue)
...
+        dev->set_retries(10);
+
+        // enable the led
+        dev->write_register(TOSHIBA_LED_ENABLE, 0x03);
+
+        /* 4-bit for each color */
+        uint8_t val[4] = { 
+            TOSHIBA_LED_PWM0, 
+            (uint8_t)(blue >> 4),
+            (uint8_t)(green / 16),
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:606]: I see in the code where you copied this example from. I'm puzzled why it uses >>4 and then /16. I'd much rather see >>4.

File: Tools/AP_Periph/can.cpp [hunk 4]
```diff
@@ -583,6 +583,32 @@ static void set_rgb_led(uint8_t red, uint8_t green, uint8_t blue)
...
+#define TOSHIBA_LED_PWM0    0x01    // pwm0 register
+#define TOSHIBA_LED_ENABLE  0x04    // enable register
+#define TOSHIBA_LED_I2C_ADDR 0x55    // default I2C bus address
+
+        static AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev;
+        if (!dev) {
+            dev = std::move(hal.i2c_mgr->get_device(0, TOSHIBA_LED_I2C_ADDR));
+        }
+        WITH_SEMAPHORE(dev->get_semaphore());
+        dev->set_retries(10);
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:597]: lets reduce this to 1

``dev->set_retries(1);``

In a different PR I'd like to change the other two drivers to (1) as well

File: Tools/AP_Periph/can.cpp [hunk 5]
```diff
@@ -583,6 +583,32 @@ static void set_rgb_led(uint8_t red, uint8_t green, uint8_t blue)
...
+
+        // enable the led
+        dev->write_register(TOSHIBA_LED_ENABLE, 0x03);
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:600]: Does this really need to be called on every led update? Seems like excessive i2c transactions. This should go either up into where dev gets set or in a ```static bool run_once;`` thing right there.

**Review by magicrub — CHANGES_REQUESTED** [review]

**Comment by magicrub** [comment]
This doesn't effect the bootloader so lets not update that in this PR if we don't need to.

File: Tools/AP_Periph/can.cpp (hunk 5 above)
Comment (davidbitton) [reply Tools/AP_Periph/can.cpp:600]: in AP_Notify, there is an init sequence that sets that register, then retries drops to 1 within the same init. `cann.cpp` is all static functions. I looked at moving that inside the lazy load, but was unaware of the consequences if having it set to 10. I will move.

**Comment by davidbitton** [comment]
If you do not mind, please do make those changes. Then I can use it as a template. Thanks.

File: Tools/AP_Periph/can.cpp (hunk 4 above)
Comment (davidbitton) [reply Tools/AP_Periph/can.cpp:597]: Maybe not. The `AP_Notify` code had it at 10, but in an init sequence; then it becomes 1 there on. I wasn't seeing success at 1, so I put back to 10, and it blinked. See can.cpp comments.

**Comment by magicrub** [comment]
sure thing

**Review by davidbitton — COMMENTED** [review]
ok.

**Comment by magicrub** [comment]
@davidbitton are you going to make that change we discussed on discord or would you like me to? I haven't done anything because you said you had additional changes on your end..

**Review by magicrub — APPROVED** [review]

**Comment by magicrub** [comment]
@davidbitton can you do another quick test on this? I think it's ready to go in now but I made some tweaks that needs a test.

**Pull request merged**